## написать пайплайн из трех шагов
## 1. взять риды и запуситить fastqc
## 2. выровнять риды на референс с помощью minimap2
## 3. посчитать покрытие с помощью bedtools genomecov


assembly_fasta = "./test/assembly_with_Ns.fasta.gz"

rule all:
    input:
        expand("results/{sample}/fastqc/{sample}_fastqc.html", 
               sample=["sample_1_reads", "sample_2_reads", "sample_3_reads", "hifi_reads"]),
        expand("results/{sample}/coverage/{sample}_coverage.txt", 
               sample=["sample_1_reads", "sample_2_reads", "sample_3_reads", "hifi_reads"])

rule fastqc:
    input:
        reads= "./test/{sample}.fastq"
    output:
        html="results/{sample}/fastqc/{sample}_fastqc.html",
        zip="results/{sample}/fastqc/{sample}_fastqc.zip"
    conda:
        "envs/fastqc_env.yaml"
    shell:
        """
        fastqc {input.reads} --outdir=results/{wildcards.sample}/fastqc
        """

rule align_minimap2:
    input:
        reads="./test/{sample}.fastq",
        ref=assembly_fasta
    threads: max(2, workflow.cores)
    output:
        bam="results/{sample}/alignment/{sample}_to_assembly.bam"
    conda:
        "envs/minimap_env.yaml"
    shell:
        """
        minimap2 -a {input.ref} {input.reads} | samtools sort -o {output.bam}
        samtools index {output.bam}
        """

rule coverage_bedtools:
    input:
        bam=rules.align_minimap2.output.bam
    output:
        cov="results/{sample}/coverage/{sample}_coverage.txt"
    conda:
        "envs/minimap_env.yaml"
    shell:
        """
        bedtools genomecov -ibam {input.bam} > {output.cov}
        """